### ISSUE TITLE
Implement nlm

### ISSUE THREAD

**Issue by mardukbp** [oracle/fastr#91]
The [oSCR](https://github.com/jaroyle/oSCR/) package is written using only R and the calculations can take hours or even days. Therefore, I thought it would be great to try it with FastR. However, it uses the function `nlm`, which is not implemented yet.

Is it doable in a short time? And if so, how should I proceed?

EDIT: There is an example that calls `nlm` in `com.oracle.truffle.r.native/gnur/tests/src/Examples/stats-Ex.R`.

**Comment by steve-s** [comment]
Hello Marduk, 

thank you for your interest in FastR. 

Yes, `nlm` is in our TODO list. It is a C function in the `stats` package (in file [`src/optimize.c`](https://github.com/wch/r-source/blob/trunk/src/library/stats/src/optimize.c)) called via the `.External2` built-in. FastR should support `.External2` and most of the R extensions API defined in [`Rinternals.h`](https://github.com/wch/r-source/blob/trunk/src/include/Rinternals.h) just fine. The issue is that `stats`, being a base package, also [includes `Defn.h`](https://github.com/wch/r-source/blob/trunk/src/library/stats/src/optimize.c#L27) (and potentially uses some functions defined there). Note it also [includes `R_ext/Applic.h`](https://github.com/wch/r-source/blob/trunk/src/library/stats/src/optim.c#L25), but FastR also provides this header and most functions defined there.

The approach to porting this to FastR is to:

- copy&paste the vanilla [`optimize.c`](https://github.com/wch/r-source/blob/trunk/src/library/stats/src/optimize.c) file from GNU-R to a directory where we in FastR maintain taken over and potentially patched GNU-R sources: [`com.oracle.truffle.r.native/gnur/patch/src/library/stats/src`](https://github.com/oracle/fastr/tree/master/com.oracle.truffle.r.native/gnur/patch/src/library/stats/src).

- remove the `#include <Defn.h>` (and maybe other unrelated functions defined in that file to reduce the noise) and compile FastR (`mx build`, the dependency analysis in the makefiles is bit quirky, you may have to `mx clean` if the changes are not picked up). You will get missing symbols errors (functions and other stuff from `Defn.h`). Now the difficult part comes: rewrite the code so that is doesn't use them or re-define them as helper functions in that file, but only using functions from `Rinternals.h`. Ideally stay as close to the original code as possible so that it's simple to back port upstream changes.

- you may have to copy&paste some more code from [GNU-R's appl directory](https://github.com/wch/r-source/tree/trunk/src/appl) to [FastR's appl directory](https://github.com/oracle/fastr/tree/master/com.oracle.truffle.r.native/gnur/patch/src/appl).

- [remove the error](https://github.com/oracle/fastr/blob/master/com.oracle.truffle.r.nodes.builtin/src/com/oracle/truffle/r/nodes/builtin/base/foreign/CallAndExternalFunctions.java#L566) and let FastR just proceed to call the `nlm` C function.

If you do futher experiment with oSCR and FastR please let us know how it went.

**Comment by mardukbp** [comment]
Thanks a lot Stepan for your detailed reply! So I added `optimize.c` without modification and `appl/uncmin.c` (which implements the actual optimization algorithm) removing only code related to printing. I ran a short example from oSCR and... FastR takes twice as long as R!

Why is that? How can the performance be improved? I was expecting a speedup not a slowdown.

Note: I compiled FastR with `mx build` using Oracle's JDK 1.8 on macOS 10.12.

**Comment by steve-s** [comment]
I assume you are running dev build of FastR? It's fine to use stock JDK for development, but in order to allow Graal to optimize FastR's code, you need JVMCI enabled JDK (can be downloaded from [here](https://github.com/graalvm/openjdk8-jvmci-builder/releases)). Then you need to run FastR with

```
mx --dynamicimports graal/compiler R
```

the `--dynamicimports graal/compiler` will enable the Graal compiler (your `JAVA_HOME` environment variable must point to the JVMCI enabled JDK). Note that all this is already configured for you if you use GraalVM distribution, but it has to be done manually for dev build.

FastR (and Graal) compiles your R program while it's running, so it usually takes a bit longer to warm-up to the best performance. I suggest you put your benchmark in a loop and check if the performance improves with further iterations. To get a log of what R functions are currently being compiled you can run FastR like so:

```
mx --J @'-Dgraal.TraceTruffleCompilationDetails=true' --dynamicimports graal/compiler R
```

(in GraalVM it would be `./bin/R --vm.Dgraal.TraceTruffleCompilationDetails=true ...`). The logging should cease at some point -- that's when all the R functions that the code is using have been compiled. If it never stops, it's a bug on our side.

You can also run FastR like so:

```
mx --cp-sfx ../graal/tools/mxbuild/dists/jdk1.8/truffle-profiler.jar --dynamicimports graal/compiler R
```

in which case you can use the VisualVM tool from GraalVM (`bin/jvisualvm`) and its feature called Graal Sampler to profile the benchmark (CPU sampling) -- find the FastR process in the left panel, open it and switch to "Graal Sampler" tab.

By default, C code is run natively even in FastR, which means that if the hot spot of your benchmark is `nlm` itself, you should eventually get to a similar speed as GNU-R. Both FastR and GNU-R would be spending most of the time in the very same C code. You can switch FastR to a mode where it executes native code via LLVM bitcode directly executed inside JVM (in that case, the Graal Sampler should show you even stack traces and hot spots of the C code):

```
mx --dynamicimports graal/compiler R --R.BackEndLLVM
```

**Comment by mardukbp** [comment]
Thank you so much Stepan! Now the oSCR example runs faster than GNU-R, as expected. And using the VisualVM tool I was able to identify the hot spot in the R code. This is awesome for R development! Now I am going to optimize that function using the hints from [Advanced R](http://adv-r.had.co.nz/Performance.html).

I also learned about [pqR - a "pretty quick" version of R](https://github.com/radfordneal/pqR). The oSCR example runs in less than 2 min, whereas in FastR it takes 3 min (after warming it up with a first run). The author claims that GNU-R is not willing to incorporate his [optimizations](https://github.com/radfordneal/pqR/issues/36#issuecomment-478802903). But it would be great if FastR did. His knowledge and expertise in optimizing R would be very beneficial for FastR and the R community.

I am going to experiment with more oSCR examples, write up what I learned doing this to help other scientists get started using FastR and prepare the pull request that solves this issue.

It would be great if I could modify the R sources in `gnur/patch/src` and recompile only that instead of having to `mx clean && mx build` every time. It is too time consuming. How can I achieve that?

**Comment by steve-s** [comment]
> I also learned about pqR - a "pretty quick" version of R. The oSCR example runs in less than 2 min, whereas in FastR it takes 3 min (after warming it up with a first run).

it would be interesting to know where pqR gains the speedup. Whether there is some particular operation or builtin function that is significantly faster on pqR.

> The author claims that GNU-R is not willing to incorporate his optimizations. But it would be great if FastR did. His knowledge and expertise in optimizing R would be very beneficial for FastR and the R community.

I wouldn't be against integrating some of these. Few notes on each of them:

- Support for deferred evaluation: we plan to work on this. Simple cases, like individual expressions, should be relatively simple. However, I am not sure how much is this transferable from pqR to FastR.

- A cleaner and faster parser: this could be quite transferable (basically rewrite of the C code -> Java). It won't help with peak performance, but it could make a difference for the start-up. Moreover, the interface between the parser and rest of FastR is relatively simple, so this could be nice isolated and mostly self-contained change.

- A "variant result" mechanism: I think this should be already done by the optimizing Graal compiler for FastR.

- Fast matrix multiplication using SIMD instructions and multiple cores: rewriting the (key) builtin functions to use multiple cores is on our TODO list. You cannot use SIMD instructions explicitly in Java but the Graal compiler should generate them for specific code patterns (e.g., simple loop over an array doing specific operation) so the important part would be to check that Graal actually vectorizes the important loops in the Java code.

> I am going to experiment with more oSCR examples, write up what I learned doing this to help other scientists get started using FastR and prepare the pull request that solves this issue.

great!

> It would be great if I could modify the R sources in gnur/patch/src and recompile only that instead of having to mx clean && mx build every time. It is too time consuming. How can I achieve that?

What should also work is removing the `fastr/com.oracle.truffle.r.native/gnur/patch-build` directory and `fastr/com.oracle.truffle.r.native/common.done` and then just `mx build`.

[This](https://github.com/oracle/fastr/blob/master/com.oracle.truffle.r.native/gnur/Makefile.gnur#L150) is where we copy the files from `patch` to `patch-build`. [This](https://github.com/oracle/fastr/blob/master/com.oracle.truffle.r.native/Makefile#L81) is where we build the base packages from `patch-build. [This](https://github.com/oracle/fastr/blob/master/com.oracle.truffle.r.native/fficall/src/common/Makefile) is where we build the other sources from `patch-build/main`.

This part of our built process is a bit messy and could use some improvements :)

**Comment by steve-s** [comment]
Hello Marduk,

finally, your OCA signature is in our system. It would be awesome if you could also contribute your `nlm` port.

**Comment by mardukbp** [comment]
Hi Stepan, I tried another oSCR example and when it calls `nlm` FastR throws several exceptions. The log is too long. I uploaded it as a [gist](https://gist.github.com/mardukbp/c309505f97340a041f0a342a12b567aa).

**Comment by steve-s** [comment]
Hello Marduk, 

these [two](https://github.com/oracle/fastr/blob/master/com.oracle.truffle.r.nodes.builtin/src/com/oracle/truffle/r/nodes/builtin/base/MatMult.java#L121) [methods](https://github.com/oracle/fastr/blob/master/com.oracle.truffle.r.nodes.builtin/src/com/oracle/truffle/r/nodes/builtin/base/MatMult.java#L131) handle the situation where one of the arguments of `%*%` is matrix that has one of its dimensions == 0, e.g., in `matrix(0, nrow=0, ncol=3) %*% matrix(0, nrow=3, ncol=1)`, but these methods fail to handle the possibility that the other argument that doesn't need to pass the `hasZeroDim` guard could be a normal vector and not a matrix (hence the `NullPointerException` because the dimensions retrieved via `getBDimsNode.getDimensions(...)` are `null`). Simple reproducing example is `matrix(0, nrow=0, ncol=1) %*% c(1,2,3)`.

Please let us know if you would like to fix this as part of your effort otherwise we will put it in our todo list.

**Comment by mardukbp** [comment]
Thank you Stepan. I will have a look. Speaking of issues with dimensions, in order to obtain this error I had to overcome another error first. The R `stats::dist` function complained that its argument is not a square matrix. After setting the storage mode of its argument to "double" the problem was gone. Any hints?

**Comment by steve-s** [comment]
Hello Marduk, 

do you know if this happens in the R wrapper or in the actual C function that the wrapper calls via `.Call(C_dist, .....)`? We ported some of the simpler C functions from the `stats` package to Java and `C_dist` is one of them. My first guess is that the error you get is [caused by this code](https://github.com/oracle/fastr/blob/master/com.oracle.truffle.r.library/src/com/oracle/truffle/r/library/stats/Cdist.java#L71) and it is a bug or incomplete implementation of `C_dist` in FastR.

**Comment by mardukbp** [comment]
Thank you Stepan, I will look into it. These days I am a bit busy with job applications, but I am still on it. 

During the weekend I tried [Renjin](https://renjin.org), the JVM-based R interpreter. I am happy to report that FastR beats Renjin in its own [example benchmark](http://docs.renjin.org/en/latest/package/index.html#usage) by three orders of magnitude. What I found more interesting though is that they compile R packages, including C and Fortran source code, to Java bytecode. They achieve this feat using [GCC-Bridge](https://github.com/bedatadriven/renjin/tree/master/tools/gcc-bridge), which includes a compiler from GCC's IR Gimple to JVM bytecode. They already have lots of R packages [in Maven](https://mvnrepository.com/search?q=renjin).

I believe it would be amazing for FastR to take advantage of this strategy to automatically incorporate the large majority of the CRAN packages. What do you think?

**Comment by mardukbp** [comment]
I just submitted the pull request #100 that should solve this issue.

**Comment by steve-s** [comment]
Thank you Marduk! I am out of the office this week so I will be able to review and merge your PRs next week.

**Comment by steve-s** [comment]
Fixed in ad8035c4a1d8385ce786ab373088c391d015c6ba
